File: internal/service/post.go
package service

import (
	"context"
	"errors"

	"AI-Study-Community/internal/dao"
	"AI-Study-Community/internal/model/do"
	"AI-Study-Community/internal/model/entity"

	"github.com/gogf/gf/v2/frame/g"
)

// Post is the service layer for posts
type Post struct {
}

// PostService is the singleton instance of Post service
var PostService = &Post{}

// ErrPostNotOwned is returned when a user tries to modify a post they don't own
var ErrPostNotOwned = errors.New("post not owned by user")

// Create creates a new post.
// It validates the input and calls the DAO layer to insert the post.
// Returns the created post ID.
func (s *Post) Create(ctx context.Context, req *do.PostCreateReq) (*do.PostCreateResp, error) {
	// Call DAO layer to insert the post
	id, err := dao.PostDao.Insert(ctx, req)
	if err != nil {
		return nil, err
	}

	return &do.PostCreateResp{
		Id: id,
	}, nil
}

// GetDetail retrieves a single post by ID.
// It calls the DAO layer to fetch the post details.
// Returns the post entity or nil if not found.
func (s *Post) GetDetail(ctx context.Context, id uint64) (*entity.Post, error) {
	return dao.PostDao.GetOne(ctx, id)
}

// Update updates an existing post by ID.
// It validates the input and calls the DAO layer to update the post.
// Returns true if the update was successful.
func (s *Post) Update(ctx context.Context, req *do.PostUpdateReq) (bool, error) {
	rowsAffected, err := dao.PostDao.Update(ctx, req)
	if err != nil {
		return false, err
	}
	return rowsAffected > 0, nil
}

// Delete soft-deletes a post by ID.
// It calls the DAO layer to perform the deletion.
// Returns true if the deletion was successful.
func (s *Post) Delete(ctx context.Context, id uint64,currentUserId uint64) (bool, error) {


	post, err := dao.PostDao.GetOne(ctx, id)
	if err != nil {
		return false, err
	}

	if post.AuthorId != currentUserId {
		g.Log().Errorf(ctx, "Delete post permission denied: post.AuthorId=%d, currentUserId=%d", post.AuthorId, currentUserId)
		return false, ErrPostNotOwned
	}	

	rowsAffected, err := dao.PostDao.Delete(ctx, id)
	if err != nil {
		return false, err
	}
	return rowsAffected > 0, nil
}

// GetPageList retrieves a paginated list of posts with optional filters.
// It calls the DAO layer to fetch the post list and total count.
// Returns the paginated post list response.
func (s *Post) GetPageList(ctx context.Context, req *do.PostGetPageListReq) (*do.PostListResp, error) {
	// Convert to DAO request format
	daoReq := &do.PostGetListReq{
		Page:     int(req.Page),
		PageSize: int(req.PageSize),
		AuthorId: req.AuthorId,
		Tags:     req.Tags,
		Status:   req.Status,
		Keyword:  req.Keyword,
	}

	// Call DAO layer to get paginated list
	list, total, err := dao.PostDao.GetList(ctx, daoReq)
	if err != nil {
		return nil, err
	}

	// Convert entity posts to response DTOs
	postList := make([]*do.PostResp, 0, len(list))
	for _, post := range list {
		postList = append(postList, &do.PostResp{
			Id:           post.Id,
			Title:        post.Title,
			Content:      post.Content,
			AuthorId:     post.AuthorId,
			Status:       post.Status,
			ViewCount:    post.ViewCount,
			LikeCount:    post.LikeCount,
			CommentCount: post.CommentCount,
			Tags:         post.Tags,
			CoverImage:   post.CoverImage,
			IsDeleted:    post.IsDeleted,
			CreatedAt:    post.CreatedAt,
			UpdatedAt:    post.UpdatedAt,
		})
	}

	return &do.PostListResp{
		Total:    uint64(total),
		Page:     req.Page,
		PageSize: req.PageSize,
		List:     postList,
	}, nil
}

File: frontend/src/api/post.js
/**
 * Post API - 帖子相关接口封装
 * 对应后端 API:
 * - POST /api/post/create   - 发布帖子
 * - POST /api/post/update   - 更新帖子
 * - POST /api/post/delete   - 删除帖子
 * - GET  /api/post/detail   - 查询帖子详情
 * - GET  /api/post/list     - 分页查询帖子列表
 */
import request from './request.js'

/**
 * 发布帖子
 * @param {Object} data - 帖子数据
 * @param {string} data.title - 帖子标题
 * @param {string} data.content - 帖子内容
 * @param {number} data.authorId - 作者ID
 * @param {string} [data.tags] - 标签（逗号分隔）
 * @param {string} [data.coverImage] - 封面图片URL
 * @returns {Promise<{id: number}>} 返回新帖子ID
 */
export function createPost(data) {
  return request({
    url: '/api/post/create',
    method: 'POST',
    data,
  })
}

/**
 * 更新帖子
 * @param {Object} data - 帖子数据
 * @param {number} data.id - 帖子ID
 * @param {string} data.title - 帖子标题
 * @param {string} data.content - 帖子内容
 * @param {number} data.authorId - 作者ID
 * @param {string} [data.tags] - 标签（逗号分隔）
 * @param {string} [data.coverImage] - 封面图片URL
 * @returns {Promise<{id: number}>} 返回更新的帖子ID
 */
export function updatePost(data) {
  return request({
    url: '/api/post/update',
    method: 'POST',
    data,
  })
}

/**
 * 删除帖子
 * @param {number} id - 帖子ID
 * @returns {Promise<{success: boolean}>} 返回删除是否成功
 */
export function deletePost(id) {
  return request({
    url: '/api/post/delete',
    method: 'POST',
    data: { id },
  })
}

/**
 * 查询帖子详情
 * @param {number} id - 帖子ID
 * @returns {Promise<Object>} 返回帖子详情
 */
export function getPostDetail(id) {
  return request({
    url: '/api/post/detail',
    method: 'GET',
    params: { id },
  })
}

/**
 * 分页查询帖子列表
 * @param {Object} params - 查询参数
 * @param {number} [params.page=1] - 页码
 * @param {number} [params.pageSize=10] - 每页数量
 * @param {number} [params.authorId] - 作者ID（可选）
 * @param {string} [params.tags] - 标签筛选（可选）
 * @param {number} [params.status] - 状态筛选（可选）：0-已删除 1-正常 2-置顶 3-推荐
 * @param {string} [params.keyword] - 关键词搜索（可选）
 * @returns {Promise<Object>} 返回分页列表 { total, page, pageSize, list }
 */
export function getPostList(params) {
  return request({
    url: '/api/post/list',
    method: 'GET',
    params,
  })
}

// 导出默认对象，方便按需引入
export default {
  createPost,
  updatePost,
  deletePost,
  getPostDetail,
  getPostList,
}
